src/box/memtx_engine.cc

```diff
@@ -107,19 +107,58 @@ memtx_alloc_init(void)
 	memtx_tuple_new_raw = memtx_tuple_new_raw_impl<ALLOC>;
 }
 
+/*
+ * Usually during final recovery, when .xlog files are loaded and applied,
+ * a special memtx_space_replace_primary_key function is used for faster
+ * recovery. The function only updates primary key while all other indexes
+ * are scheduled to build at the end of recovery.
+ *
+ * This approach doesn't work with MVCC engine and synchro spaces: when
+ * transactions appear in xlog and they may or may not be committed, we
+ * should process them in normal mode throughout all indexes in order to
+ * leave transaction history in each index.
+ *
+ * So all space indexes are built early in this case.
+ */
+static bool
+memtx_build_spaces_early()
+{
+	return memtx_tx_manager_use_mvcc_engine;
+}
+
 static int
 memtx_end_build_primary_key(struct space *space, void *param)
```

Comment:
Looks like the `param` argument type can be changed from `void *` to `memtx_engine *` here and in other places.

Reply:
Done for the `memtx_end_build_primary_key` function.

`memtx_build_secondary_keys` and `memtx_begin_empty_space_final_recovery` are used with `space_foreach`, so the `void *` param has to be kept there.